pkg/controller/daemon/update.go

```diff
@@ -295,23 +300,56 @@ func (dsc *DaemonSetsController) controlledHistories(ds *extensions.DaemonSet) (
 }
 
 // Match check if ds template is semantically equal to the template stored in history
-func Match(template *v1.PodTemplateSpec, history *apps.ControllerRevision) (bool, error) {
-	t, err := DecodeHistory(history)
-	return apiequality.Semantic.DeepEqual(template, t), err
+func Match(ds *extensions.DaemonSet, history *apps.ControllerRevision) (bool, error) {
+	patch, err := getPatch(ds)
+	if err != nil {
+		return false, err
+	}
+	return bytes.Equal(patch, history.Data.Raw), nil
 }
 
-func DecodeHistory(history *apps.ControllerRevision) (*v1.PodTemplateSpec, error) {
-	template := v1.PodTemplateSpec{}
-	err := json.Unmarshal(history.Data.Raw, &template)
-	return &template, err
+// ApplyHistory returns a specific revision of DaemonSet by applying the given history to a copy of the given DaemonSet
+func ApplyHistory(ds *extensions.DaemonSet, history *apps.ControllerRevision) (*extensions.DaemonSet, error) {
+	obj, err := scheme.Scheme.DeepCopy(ds)
+	if err != nil {
+		return nil, err
+	}
+	clone := obj.(*extensions.DaemonSet)
+	patched, err := strategicpatch.StrategicMergePatch([]byte(runtime.EncodeOrDie(patchCodec, clone)), history.Data.Raw, clone)
```

Comment:
Do you really need a patchCodec that knows the scheme? It looks like you only need a json encoder.